pkg/frontend/querymiddleware/roundtrip_test.go

```diff
@@ -153,20 +153,47 @@ func TestInstantTripperware(t *testing.T) {
 		})
 	})
 
-	queryClient, err := api.NewClient(api.Config{Address: "http://localhost", RoundTripper: tw(rt)})
-	require.NoError(t, err)
-	api := v1.NewAPI(queryClient)
+	tripper := tw(rt)
+
+	t.Run("specific time happy case", func(t *testing.T) {
+		queryClient, err := api.NewClient(api.Config{Address: "http://localhost", RoundTripper: tripper})
+		require.NoError(t, err)
+		api := v1.NewAPI(queryClient)
 
-	t.Run("happy case roundtrip", func(t *testing.T) {
+		ts := time.Date(2021, 1, 2, 3, 4, 5, 0, time.UTC)
 		res, _, err := api.Query(ctx, `sum(increase(we_dont_care_about_this[1h])) by (foo)`, ts)
 		require.NoError(t, err)
 		require.Equal(t, model.Vector{
 			{Metric: model.Metric{"foo": "bar"}, Timestamp: model.TimeFromUnixNano(ts.UnixNano()), Value: totalShards},
 		}, res)
 	})
 
-	t.Run("default time param", func(t *testing.T) {
-		res, _, err := api.Query(ctx, `sum(increase(we_dont_care_about_this[1h])) by (foo)`, time.Now())
+	t.Run("default time param happy case", func(t *testing.T) {
+		queryClient, err := api.NewClient(api.Config{Address: "http://localhost", RoundTripper: tripper})
+		require.NoError(t, err)
+		api := v1.NewAPI(queryClient)
+
+		res, _, err := api.Query(ctx, `sum(increase(we_dont_care_about_this[1h])) by (foo)`, time.Time{})
+		require.NoError(t, err)
+		require.IsType(t, model.Vector{}, res)
+		require.NotEmpty(t, res.(model.Vector))
+
+		resultTime := res.(model.Vector)[0].Timestamp.Time()
+		require.InDelta(t, time.Now().Unix(), resultTime.Unix(), 1)
+	})
+
+	t.Run("default time param with form being already parsed", func(t *testing.T) {
```

Comment:
WDYT about adding a test case showing request body taking precedence over the query string?

Reply:
Added a test here: https://github.com/grafana/mimir/pull/3588/commits/1330348ce4c34a28e383b2a54a63edc409bab535